File: Ch02/challenge.cpp
#include <iostream>
#include <map>
#include <ctime>
#include <cstdlib>

/* Problemas de implementação: o cache é compartilhado para todas as memoizaçõe */

using namespace std;

auto const memoize = [](auto f) {
	return [f](unsigned int x) {
		static map<int, int> cache;
		if (cache.find(x)==cache.end()) cache[x] = f(x);
		return cache[x];
	};
};

unsigned int fibo(unsigned int n) {
	if (n==0 || n==1) return n;
	return fibo(n-1) + fibo(n-2);
}

unsigned int report_time(auto f, auto n) {
	time_t t1, t2;
	unsigned int x;

	t1 = time(NULL);
	x = f(n);
	t2 = time(NULL);

	cout << "Result: " << x << " in " << t2-t1 << " secs." << endl;

	return x;
}

unsigned int rand_range(unsigned int x) {
	return rand() % x;
}

unsigned int rand_seed(unsigned int x) {
	srand(x);
	return rand();
}

int main() {

	auto mfibo = memoize(fibo);
	auto mrange = memoize(rand_range);
	auto mseed = memoize(rand_seed);

	report_time(fibo, 45);
	report_time(mfibo, 45);
	report_time(mfibo, 45);

	cout << rand_range(100) << endl;
	cout << rand_range(100) << endl;
	cout << mrange(100) << endl;
	cout << mrange(100) << endl;

	cout << rand_seed(120) << endl;
	cout << rand_seed(120) << endl;
	cout << mseed(120) << endl;
	cout << mseed(120) << endl;
	
	return 0;
}

File: Ch02/challenge.py
import time
import random

def memoize(f):
    mem = {}
    def memf(x):
        if x not in mem:
            mem[x] = f(x)
        return mem[x]
    return memf

def fibo(n):
    if n == 0: return 0
    if n == 1: return 1
    return fibo(n-1) + fibo(n-2)

def rand_seed(seed):
    random.seed(seed)
    return random.random()

f = memoize(fibo)
t1 = time.time()
print(f(30))
t2 = time.time()
print(f(30))
t3 = time.time()

diff1 = t2-t1
diff2 = t3-t2
print(f"First call: {diff1}, Second call: {diff2}")

g = memoize(random.randrange)
m1, m2, m3 = g(100), g(100), g(200)
n1, n2, n3 = random.randrange(100), random.randrange(100), random.randrange(200)
print(f"Random numbers: {m1}, {m2}, {m3}")
print(f"Memoized random numbers: {n1}, {n2}, {n3}")

g = memoize(rand_seed)
m1, m2, m3 = g(100), g(100), g(200)
n1, n2, n3 = rand_seed(100), rand_seed(100), rand_seed(200)
print(f"Random numbers: {m1}, {m2}, {m3}")
print(f"Memoized random numbers: {n1}, {n2}, {n3}")
